src/React/Basic.purs

```diff
@@ -32,25 +29,57 @@ foreign import react_
 -- | constructed using the helper functions provided by the `React.Basic.DOM`
 -- | module (and re-exported here).
 react
-  :: forall props state
-   . { initialState :: state
-     , receiveProps :: props -> state -> (state -> Eff (react :: ReactFX) Unit) -> Eff (react :: ReactFX) Unit
-     , render :: props -> state -> (state -> Eff (react :: ReactFX) Unit) -> JSX
+  :: forall props state fx
+   . { displayName :: String
+     , initialState :: state
+     , receiveProps :: props -> state -> (SetState state fx) -> Eff (react :: ReactFX | fx) Unit
+     , render :: props -> state -> (SetState state fx) -> JSX
+     }
+  -> props
+  -> JSX
+react = createElement <<< reactComponent
+
+-- | Create a React component from a _specification_ of that component, without
+-- | pre-applying it to `createElement`. Use this function for more standard
+-- | JavaScript React interop.
+reactComponent
+  :: forall props state fx
+   . { displayName :: String
+     , initialState :: state
+     , receiveProps :: props -> state -> (SetState state fx) -> Eff (react :: ReactFX | fx) Unit
+     , render :: props -> state -> (SetState state fx) -> JSX
      }
   -> ReactComponent props
-react { initialState, receiveProps, render } =
-  react_
-    { initialState
+reactComponent { displayName, initialState, receiveProps, render } =
+  component_
+    { displayName
+    , initialState
     , receiveProps: mkEffFn3 receiveProps
     , render: mkFn3 render
     }
 
-foreign import component_ :: forall props. Fn2 (ReactComponent props) props JSX
+-- | SetState uses an update function to modify the current state.
+type SetState state fx = (state -> state) -> Eff (react :: ReactFX | fx) Unit
 
--- | Create a `JSX` node from another React component, by providing the props.
-component
+-- | Create a `JSX` node from a React component, by providing the props.
+createElement
   :: forall props
    . ReactComponent props
   -> props
   -> JSX
-component = runFn2 component_
+createElement = runFn2 createElement_
+
+foreign import keyed :: Array { key :: String, child :: JSX } -> JSX
```

Comment:
What about using a homogeneous record for this?

Reply:
It's much harder to construct in PS and slightly harder to destruct in JS.  This is for dynamic lists, where you're mapping over some data to produce JSX.

Reply:
I should add docs to this